src/detect.c

```diff
@@ -1296,6 +1296,46 @@ static inline void StoreDetectFlags(DetectTransaction *tx, const uint8_t flow_fl
     }
 }
 
+static inline RuleMatchCandidateMergeSorted(DetectEngineThreadCtx *det_ctx, uint32_t j, uint32_t k)
```

Comment:
I'm having trouble understanding this. When we get called we have have 0-N entries in `tx_candidates` from prefilter. These are already ordered. The candidates coming from the `match_array` are merged into this, and the result has to be a sorted list by the candidates `s->num` (internal sig id). I guess I'm not seeing how we insert sort things into the already existing list.

Reply:
Is this a matter of where this has been called, together with the fact that we still have the qsort called later on, (and we're not taking `num` into consideration when doing the ordering comparison), or are these factors irrelevant here?

Reply:
>  with the fact that we still have the qsort called later on

The point is to avoid the call to qsort (in the case we do not have match candidates added after by stored flags) cf removal of `do_sort = (array_idx > x); // sort if match added anything`

Reply:
> I'm having trouble understanding this

Will put more comments.

Quick and dirty answer :
This is basically merging two sorted lists.
The trick is do it in place, without shifting all the elements.
So we start from the end (if we start from the beginning, and the new element should be first, we would have to shift all j elements already in place)